conversion/data_from_database.go

```diff
@@ -64,6 +69,49 @@ func (dd *DataFromDatabaseImpl) dataFromDatabaseForDMSMigration() (*writer.Batch
 // 4. Launch the stream for the physical shard
 // 5. Perform streaming migration via dataflow
 func (dd *DataFromDatabaseImpl) dataFromDatabaseForDataflowMigration(targetProfile profiles.TargetProfile, ctx context.Context, sourceProfile profiles.SourceProfile, conv *internal.Conv, is common.InfoSchemaInterface) (*writer.BatchWriter, error) {
+	// Fetch Spanner Region
+	if conv.SpRegion == "" {
+		spInstanceAdmin, err := spinstanceadmin.NewInstanceAdminClientImpl(ctx)
+		if err != nil {
+			return nil, fmt.Errorf("unable to fetch Spanner Region for resource creation: %v", err)
+		}
+		spAcc := spanneraccessor.SpannerAccessorImpl{}
+		spannerRegion, err := spAcc.GetSpannerLeaderLocation(ctx, spInstanceAdmin, "projects/" + targetProfile.Conn.Sp.Project+ "/instances/" + targetProfile.Conn.Sp.Instance)
+		if err != nil {
+			return nil, fmt.Errorf("unable to fetch Spanner Region for resource creation: %v", err)
+		}
+		conv.SpRegion = spannerRegion
+	}
+
+	storageClient, err := storageclient.NewStorageClientImpl(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	// Create Resources required for migration
+	if conv.ResourceValidation {
+		dsClient, err := datastreamclient.NewDatastreamClientImpl(ctx)
+		if err != nil {
+			return nil, err
+		}
+		createResources := NewValidateOrCreateResourcesImpl(&datastream_accessor.DatastreamAccessorImpl{}, dsClient, &storageaccessor.StorageAccessorImpl{}, storageClient)
+		err = createResources.ValidateOrCreateResourcesForShardedMigration(ctx, targetProfile.Conn.Sp.Project, targetProfile.Conn.Sp.Instance, false, conv.SpRegion, sourceProfile)
+		if err != nil {
+			return nil, fmt.Errorf("unable to create connection profiles: %v", err)
+		}
+	}
+
+	//Set the TmpDir from the sessionState bucket which is derived from the target connection profile
+	for _, dataShard := range sourceProfile.Config.ShardConfigurationDataflow.DataShards {
+		if dataShard.TmpDir == "" {
+		bucket, rootPath, err := GetBucket(targetProfile.Conn.Sp.Project, conv.SpRegion, dataShard.DstConnectionProfile.Name)
```

Comment:
nit/side-task: we should rename thi method to GetBucketFromDatastreamProfile or something.